Add client_credentials grant type support

alek-sys opened the pull request:
Fixes gh-51

alek-sys commented:
Hey @jgrandja, I still have a couple of to-dos:
- [ ] Support basic auth for client id / secret
- [ ] Update sample app to use JWK to sign the token. This should allow to put together an end-to-end demo

With that in mind, feel free to provide any initial feedback you have.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @author Alexey Nesterov
+ */
+public class OAuth2ClientCredentialsAuthenticationProvider implements AuthenticationProvider {
```
#72 introduced `OAuth2ClientAuthenticationProvider` and `OAuth2ClientAuthenticationToken`, which handles client authentication. The `OAuth2ClientCredentialsAuthenticationProvider` should simply assert that the current `Authentication` is `OAuth2ClientAuthenticationToken`. Take a look at `OAuth2AuthorizationCodeAuthenticationProvider` and apply the same validation check for `OAuth2ClientAuthenticationToken`.

Also, take a look at the integration test `OAuth2AuthorizationCodeGrantTests.requestWhenTokenRequestValidThenResponseIncludesCacheHeaders()` to see how the client authenticates via Basic before it gets to `OAuth2AuthorizationCodeAuthenticationProvider`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/token/TokenIssuer.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * @author Alexey Nesterov
+ */
+public interface TokenIssuer<PRINCIPAL> {
```
Please remove this as it's not needed as part of this issue. The `OAuth2ClientCredentialsAuthenticationProvider` should simply leverage `StringKeyGenerator accessTokenGenerator` the same as in `OAuth2AuthorizationCodeAuthenticationProvider`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:
```diff
@@ -0,0 +1,66 @@
+
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return authentication == OAuth2ClientCredentialsAuthenticationToken.class;
```
Change to `return OAuth2ClientCredentialsAuthenticationToken.class.isAssignableFrom(authentication)`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:
```diff
@@ -86,8 +84,10 @@
 	private final AuthenticationManager authenticationManager;
 	private final OAuth2AuthorizationService authorizationService;
 	private final RequestMatcher tokenEndpointMatcher;
-	private final Converter<HttpServletRequest, Authentication> authorizationGrantAuthenticationConverter =
```
The way I envisioned `authorizationGrantAuthenticationConverter` is to be a delegation-based implementation that simply delegates to the specific authorization grant implementation. Spring Security uses this strategy in a number of places. Take a look at `DelegatingAccessDeniedHandler` and `DelegatingOAuth2UserService` for examples.

For now, both `AuthorizationCodeAuthenticationConverter` and `ClientCredentialsAuthenticationConverter` should be both `private static class` and then create a `delegate` `Converter` that simply delegates to the appropriate impl.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/AuthorizationCodeAuthenticationConverter.java:
```diff
@@ -0,0 +1,85 @@
+import org.springframework.util.MultiValueMap;
+import org.springframework.util.StringUtils;
+
+class AuthorizationCodeAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
```
Please revert this change and keep `AuthorizationCodeAuthenticationConverter` as `private static class` in `OAuth2TokenEndpointFilter`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java:
```diff
@@ -0,0 +1,68 @@
+		}
+
+		// client_id (REQUIRED)
+		String clientId = parameters.getFirst(OAuth2ParameterNames.CLIENT_ID);
```
`OAuth2ClientAuthenticationFilter` and `OAuth2ClientAuthenticationProvider` will handle authenticating the client. Both `client_id` and `client_secret` should be removed from the validation checks. Take a look at `AuthorizationCodeAuthenticationConverter` and how it returns `clientPrincipal = SecurityContextHolder.getContext().getAuthentication()` in the `Authentication` and apply the same.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java:
```diff
@@ -0,0 +1,68 @@
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.CLIENT_SECRET);
+		}
+
+		return new OAuth2ClientCredentialsAuthenticationToken(clientId, clientSecret);
```
`scope` is an optional parameter but should also be returned in `OAuth2ClientCredentialsAuthenticationToken`

alek-sys replied:
Do you expect section [3.3. Access Token](https://tools.ietf.org/html/rfc6749#section-3.3) to be implemented as a part of this PR? E.g. returning `scope` in the response, and filtering requested scopes based on a list of scopes supported by a client?

jgrandja replied:
Yes. The `OAuth2ClientCredentialsAuthenticationProvider` should receive the "request scopes" and the provider should ensure the "request scopes"  are either a subset (or all) of the "assigned scopes". If the client requests a scope that is not in the "assigned scopes" then it should fail. If no scopes are requested then we should auto-approve. However, I'm not sure auto-approving is a good idea but we can change that later if need be.

jgrandja commented:
@alek-sys 

Regarding

> Support basic auth for client id / secret

This is already implemented in `ClientSecretBasicAuthenticationConverter`

alek-sys commented:
Thanks for feedback @jgrandja! I'm now re-working this PR based on your comments.

alek-sys commented:
Updated based on your comments.

Also, [the spec](https://tools.ietf.org/html/rfc6749#section-3.3) states:
> If the issued access token scope is different from the one requested by the client, the authorization server MUST include the "scope" response parameter to inform the client of the actual scope granted.

But currently Token Endpoint always returns `scope` in the response. Shoud l I fix it in this PR or create a separate story for this (unless there is one already)?

jgrandja commented:
The returned scope will come into play in #42 at a later point. No need to address it in this PR. But feel free to add notes to #42.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:
```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
+
+			Set<String> requestedScopes = null;
+			try {
+				requestedScopes = scopesParser.parse(scopeParameter);
```
Please move `scope` validation in `OAuth2ClientCredentialsAuthenticationProvider`

jgrandja replied:
Actually the more I thought about this one, I don't think we need to validate here. The `scope` validation should be done when the `RegisteredClient` is created - similar to `ClientRegistration` in `oauth2-client`. We know then that the configured scopes are valid and all we need to do is match the configured ones against the requested ones (as-is). So let's remove `OAuth2ScopeParser` and pass in the requested scopes as is, eg:

`new HashSet<>(Arrays.asList(StringUtils.delimitedListToStringArray(scope, " ")))`

alek-sys replied:
The validation itself (e.g. check that client has all requested scopes) is happening in the provider. The purpose of`OAuth2ScopeParser`  is only parsing `scope` string request param to a set according to RFC. It is slightly more complicated than `delimitedListToStringArray` so I decided to introduce more strict parser to cover special characters, double spaces, etc. I think it still has some value, unless you absolutely want to remove it. 

Maybe exception for cases when `scope` is not a valid string should be different then, just general BAD_REQUEST?

jgrandja replied:
I still feel that `scope` validation should happen when a client is registered, eg. when a `RegisteredClient` is created. This ensures that the `scope`(s) the client is allowed to request `RegisteredClient.scopes` are valid from the start. Even if the requested `scope` has invalid characters, it still won't match against the registered (valid) scope. So I feel this extra validation at runtime is redundant so please remove `OAuth2ScopeParser`. FYI, this validation already exists in `RegisteredClient.validateScopes()`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:
```diff
@@ -0,0 +1,59 @@
+		return converter.convert(request);
+	}
+
+	public DelegateGrantTypeAuthenticationConverter registerConverter(AuthorizationGrantType authorizationGrantType, Converter<HttpServletRequest, Authentication> authenticationConverter) {
```
Please remove this method and  provide constructor that accepts the `Map` making it immutable.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:
```diff
@@ -0,0 +1,59 @@
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.util.StringUtils;
+
+public class DelegateGrantTypeAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
```
Rename to `DelegatingAuthorizationGrantAuthenticationConverter` and make `final`

jgrandja replied:
Looks like this was missed

alek-sys replied:
Sorry for that, I missed this comment totally - for whatever reason it is not shown on `Conversations` tab, only on `Files`. So I get notifcations about your comments but have to search for them. New GitHub UI is not so polished I guess. This class is now updated.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:
```diff
@@ -0,0 +1,59 @@
+
+public class DelegateGrantTypeAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
+
+	private final HashMap<AuthorizationGrantType, Converter<HttpServletRequest, Authentication>> converters = new HashMap<>();
```
Reference `Map` instead

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:
```diff
@@ -0,0 +1,59 @@
+
+	@Override
+	public Authentication convert(HttpServletRequest request) {
+		String grantType = request.getParameter("grant_type");
```
This class should not throw any errors. It simply delegates to the correct `Converter` or returns `null` . Similar to how `DelegatingOAuth2UserService` is implemented.

jgrandja replied:
Use constants in `OAuth2ParameterNames`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:
```diff
@@ -0,0 +1,74 @@
+
+	private final StringKeyGenerator accessTokenGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder());
+
+	public OAuth2ClientCredentialsAuthenticationProvider() {
```
Please remove default constructor since it's redundant

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	public Set<String> getScopes() {
+		return scopes;
```
Prefix `this.`

alek-sys commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:
```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
+
+			OAuth2ClientAuthenticationToken clientAuthenticationToken = (OAuth2ClientAuthenticationToken) authentication;
+			if (!clientAuthenticationToken.isAuthenticated()) {
+				throwError(OAuth2ErrorCodes.INVALID_CLIENT, "client_id");
```
I see now. Converter belongs to `web` package so it shouldn't really cotain any auth-n logic. So I refactored it a bit - now converted only extracts scopes from the request plus gets current authentication from the context. The rest is done in the manager, including validating scopes, using default scopes and validating client authentication. It is much more clear now, thanks for feedback.

alek-sys commented:
Thanks @jgrandja, I've updated this PR. Now there is much better split between converted and provider.

jgrandja commented:
Great thanks @alek-sys. I will review this tomorrow.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:
```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
+	 */
+	private static class ClientCredentialsAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
+
+		private static final String SCOPE_PARAMETER_NAME = "scope";
```
Please remove `SCOPE_PARAMETER_NAME` and use `OAuth2ParameterNames.SCOPE` in `convert`

jgrandja replied:
Looks like this was missed

alek-sys commented:
@jgrandja updated based on your latest comments. Also, see my comment regarding `OAuth2ScopeParser` - I think it has some value as it validates `scope` parameter to be compliant with RFC.

jgrandja commented:
@alek-sys Can you please update the test methods to following the naming convention `**methodName**When**Condition**Then**Result**`. Take a look at `OAuth2TokenEndpointFilterTests` and `OAuth2AuthorizationCodeAuthenticationProviderTests` for examples.

Also, please add integration tests in `OAuth2ClientCredentailsGrantTests` (new). See `OAuth2AuthorizationCodeGrantTests` for example.

alek-sys commented:
@jgrandja done. Actually there was `OAuth2ClientCredentialsGrantTests` class with an integration test already. I updated it a bit to make it more like `OAuth2AuthorizationCodeGrantTests`. I also changed security configuration for it to be more realistic by overriding authentication entry point to return 401 when client is not authenticated and `client_credentials` grant is used.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.util.Assert;
+
+public class OAuth2ClientCredentialsAuthenticationToken extends AbstractAuthenticationToken {
```
- Add class-level javadoc with `@author` and `@since`
- Add `OAuth2ClientCredentialsAuthenticationTokenTests`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:
```diff
@@ -0,0 +1,57 @@
+
+public class OAuth2ClientCredentialsAuthenticationToken extends AbstractAuthenticationToken {
+
+	private final Authentication clientPrincipal;
```
Add `private static final long serialVersionUID = Version.SERIAL_VERSION_UID;`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:
```diff
@@ -0,0 +1,57 @@
+	private final Authentication clientPrincipal;
+	private final Set<String> scopes;
+
+	public OAuth2ClientCredentialsAuthenticationToken(OAuth2ClientAuthenticationToken clientPrincipal, Set<String> scopes) {
```
Use `Authentication` type instead of `OAuth2ClientAuthenticationToken `

jgrandja replied:
Update to `@Nullable Set<String> scopes` and remove the other constructor

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:
```diff
@@ -0,0 +1,57 @@
+		return this.clientPrincipal;
+	}
+
+	public Set<String> getScopes() {
```
Add `@Nullable`

alek-sys replied:
`scopes` is never null, do you mean `@NonNull`?

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:
```diff
@@ -0,0 +1,81 @@
+ * An {@link AuthenticationProvider} that converts give {@link OAuth2ClientAuthenticationToken} to a {@link OAuth2AccessTokenAuthenticationToken}
+ * using the provided token generator.
+ *
+ * @author Alexey Nesterov
```
Add `@since`

jgrandja commented on core/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProviderTest.java:
```diff
@@ -0,0 +1,113 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class OAuth2ClientCredentialsAuthenticationProviderTest {
```
Rename to `...Tests`

jgrandja commented on core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:
```diff
@@ -0,0 +1,128 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+public class OAuth2ClientCredentialsGrantTests {
```
Add `@author`

jgrandja commented on core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	@Test
+	public void requestWhenAuthorizationRequestAuthenticatedThenThenReturnTokenAndScope() throws Exception {
```
Change "AuthorizationRequest" to "TokenRequest"

jgrandja commented on core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	@Test
+	public void requestWhenAuthorizationRequestNotAuthenticatedThenUnauthorized() throws Exception {
```
Change "AuthorizationRequest" to "TokenRequest"

jgrandja commented on core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	@EnableWebSecurity
+	static class AuthorizationServerConfiguration extends WebSecurityConfigurerAdapter {
```
Please change this to:

```java
@EnableWebSecurity
@Import(OAuth2AuthorizationServerConfiguration.class)
static class AuthorizationServerConfiguration {
...
```

See `OAuth2AuthorizationCodeGrantTests`

Any changes needed to the default security should be made in `OAuth2AuthorizationServerSecurity`

alek-sys commented:
All updated based on your review @jgrandja. I only didn't add `@Nullable` since scope field is never null. Also, I noticed there is no logging anywhere in the project. Is that intentional or should I add at least debug logging to some of the components?

jgrandja commented:
Thanks for adding this feature @alek-sys !

FYI, I followed up with a polish commit as some changes were needed and I wanted to get this merged.
Let me know if you have any questions regarding the changes in the polish commit.
